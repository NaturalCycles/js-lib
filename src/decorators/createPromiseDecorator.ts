import { _getTargetMethodSignature } from './decorator.util'

export interface PromiseDecoratorCfg<RES = any, PARAMS = any> {
  decoratorName: string

  /**
   * Called BEFORE the original function.
   * If Promise is returned - it will be awaited.
   */
  beforeFn?: (r: PromiseDecoratorResp<PARAMS>) => void | Promise<void>

  /**
   * Called just AFTER the original function.
   * The output of this hook will be passed further,
   * so, pay attention to pass through (or modify) the result.
   */
  thenFn?: (r: PromiseDecoratorResp<PARAMS> & { res: RES }) => RES

  /**
   * Called on Promise.reject.
   * If `catchFn` is not present - will re-throw the error,
   * which will most likely be "unhandledRejection" (unless there's another higher-level @Decorator that will catch it).
   * If `catchFn` is present - it's responsible for handling or re-throwing the error.
   * Whatever `catchFn` returns - passed to the original output.
   */
  catchFn?: (r: PromiseDecoratorResp<PARAMS> & { err: any }) => RES

  /**
   * Fires AFTER thenFn / catchFn, like a usual Promise.finally().
   * Doesn't have access to neither res nor err (same as Promise.finally).
   */
  finallyFn?: (r: PromiseDecoratorResp<PARAMS>) => any
}

export interface PromiseDecoratorResp<PARAMS> {
  decoratorParams: PARAMS
  args: any[]
  started: number
  target: any
  key: string
  decoratorName: string
}

/**
 * @example
 * // decorators.ts
 * export const BlockingLoader = () => _createPromiseDecorator({
 *   decoratorName: 'BlockingLoader',
 *   beforeFn: () => store.commit('setBlockingLoader'),
 *   finallyFn: () => store.commit('setBlockingLoader', false),
 * })
 *
 * @experimental
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
export function _createPromiseDecorator<RES = any, PARAMS = any>(
  cfg: PromiseDecoratorCfg<RES, PARAMS>,
  decoratorParams: PARAMS = {} as any,
): MethodDecorator {
  const { decoratorName } = cfg

  return function decoratorFunction(
    target: any,
    propertyKey: string | symbol,
    pd: PropertyDescriptor,
  ): PropertyDescriptor {
    // console.log(`@Decorator.${cfg.decoratorName} called: ` + propertyKey, pd, target)
    const originalMethod = pd.value
    const key = String(propertyKey)
    const methodSignature = _getTargetMethodSignature(target, key)

    pd.value = async function (this: typeof target, ...args: any[]): Promise<any> {
      // console.log(`@${cfg.decoratorName} called inside function`)
      const started = Date.now()

      try {
        // Before function
        // console.log(`@${cfg.decoratorName} Before`)
        if (cfg.beforeFn) {
          await cfg.beforeFn({
            decoratorParams,
            args,
            key,
            target,
            decoratorName,
            started,
          })
        }

        // Original function
        let res = await originalMethod.apply(this, args)

        // console.log(`${cfg.decoratorName} After`)
        const resp: PromiseDecoratorResp<PARAMS> = {
          decoratorParams,
          args,
          key,
          target,
          decoratorName,
          started,
        }

        if (cfg.thenFn) {
          res = cfg.thenFn({
            ...resp,
            res,
          })
        }

        cfg.finallyFn?.(resp)

        return res
      } catch (err) {
        console.error(`@${decoratorName} ${methodSignature} catch:`, err)

        const resp: PromiseDecoratorResp<PARAMS> = {
          decoratorParams,
          args,
          key,
          target,
          decoratorName,
          started,
        }

        let handled = false

        if (cfg.catchFn) {
          cfg.catchFn({
            ...resp,
            err,
          })
          handled = true
        }

        cfg.finallyFn?.(resp)

        if (!handled) {
          throw err // rethrow
        }
      }
    }

    return pd
  }
}
