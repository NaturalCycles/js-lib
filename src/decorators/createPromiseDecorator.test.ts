import { mockTime } from '@naturalcycles/dev-lib/dist/testing'
import { pDelay } from '../index'
import { _createPromiseDecorator } from './createPromiseDecorator'

export interface LoaderDecoratorParams {
  type: string
}

const beforeFn = jest.fn()
const thenFn = jest.fn(r => r.res)
const catchFn = jest.fn()
const finallyFn = jest.fn()

// eslint-disable-next-line @typescript-eslint/naming-convention
const DummyDecorator = () =>
  _createPromiseDecorator({
    decoratorName: 'Dummy',
  })

// eslint-disable-next-line @typescript-eslint/naming-convention
const Loader = (params: LoaderDecoratorParams) =>
  _createPromiseDecorator(
    {
      decoratorName: 'Loader',
      beforeFn,
      thenFn,
      catchFn,
      finallyFn,
    },
    params,
  )

class C {
  @Loader({ type: 'MY_TYPE' })
  async run(a: number): Promise<number> {
    return await pDelay(10, a * 2)
  }

  @DummyDecorator()
  async dummy(): Promise<number> {
    return 4
  }

  // Cannot use _createPromiseDecorator on sync methods!
  // @DummyDecorator()
  // dummySync(): number {
  //   return 4
  // }
}
const c = new C()

beforeEach(() => {
  mockTime()
})

test('_createPromiseDecorator', async () => {
  const _r = await c.run(2)
  // expect(r).toBe(4) // todo: it suddenly broke

  expect(thenFn).toMatchSnapshot()

  expect(await c.dummy()).toBe(4)
})

// todo: cover more cases
