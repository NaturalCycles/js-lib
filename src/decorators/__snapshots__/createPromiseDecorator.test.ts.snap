// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_createPromiseDecorator 1`] = `
[MockFunction] {
  "calls": [
    [
      {
        "args": [
          2,
        ],
        "decoratorName": "Loader",
        "decoratorParams": {
          "type": "MY_TYPE",
        },
        "key": "run",
        "res": 4,
        "started": 1529539200000,
        "target": C {},
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": 4,
    },
  ],
}
`;
